components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+/// That would greatly optimize the 'sync' to on prem builder case, as we could point to that
+/// and only fetch what we don't already have.
+#[allow(clippy::too_many_arguments)]
+pub fn start<U>(ui: &mut U,
+                url: &str,
+                channel: &ChannelIdent,
+                product: &str,
+                version: &str,
+                idents: Vec<PackageIdent>,
+                target: PackageTarget,
+                fs_root_path: Option<&PathBuf>,
+                token: Option<&str>)
+                -> Result<()>
+    where U: UIWriter
+{
+    debug!("Starting download with url: {}, channel: {}, product: {}, version: {}, target: {}, \
+            fs_root_path: {:?}, token: {:?}",
+           url, channel, product, version, target, fs_root_path, token);
+
+    let key_cache_path = &cache_key_path(fs_root_path);
+    debug!("install key_cache_path: {}", key_cache_path.display());
+
+    let artifact_cache_path = &cache_artifact_path(fs_root_path);
+    debug!("install artifact_cache_path: {}",
+           artifact_cache_path.display());
+
+    // TODO we use the same root path for ssl certs as we do for the rest of the root path,
+    // We shouldn't probably override it here, as this appears to be largely for cert paths
+    let api_client = Client::new(url, product, version, fs_root_path.map(PathBuf::as_path))?;
+    let task = DownloadTask { idents,
+                              target,
+                              url,
+                              api_client,
+                              token,
+                              channel,
+                              artifact_cache_path,
+                              key_cache_path };
+
+    let downloaded_artifacts: Vec<PackageArchive> = task.execute(ui).unwrap();
+
+    debug!("Expanded package count: {}", downloaded_artifacts.len());
+
+    Ok(())
+}
+
+struct DownloadTask<'a> {
+    idents: Vec<PackageIdent>,
+    target: PackageTarget,
+    url: &'a str,
+    api_client: BoxedClient,
+    token: Option<&'a str>,
+    channel: &'a ChannelIdent,
+    /// The path to the local artifact cache (e.g., /hab/cache/artifacts)
+    artifact_cache_path: &'a Path,
+    key_cache_path: &'a Path,
+}
+
+impl<'a> DownloadTask<'a> {
+    fn execute<T>(&self, ui: &mut T) -> Result<Vec<PackageArchive>>
+        where T: UIWriter
+    {
+        // This was written intentionally with an eye towards data parallelism
+        // Any or all of these phases should naturally fit a fork-join model
+
+        ui.begin(format!("Preparing to download necessary packages for {} idents",
+                         self.idents.len()))?;
+        ui.begin(format!("Using channel {} from {}", self.channel, self.url))?;
+        ui.begin(format!("Storing in cache at {:?} ", self.artifact_cache_path))?;
+
+        // Phase 1: Expand to fully qualified deps and TDEPS
+        let expanded_idents = self.expand_sources(ui)?;
```

Comment:
I think the code will be simpler if we have the DownloadTask just do a single ident download.  Phase 1 would then get pulled outside of DownloadTask.  DownloadTask also doesn't need to return a PackageArchive (it can just return () - we only care about success/failure(err))